### PR TITLE
Various improvements for WNS notifications

### PR DESCRIPTION
**Pull request by wouterh**
Add support for:
* WNS badge notifications
* The launch argument of WNS toast notifications
* Sound in WNS toast notifications

### CODE REVIEW

**Comment by aried3r** [comment]
What do you think about leaving `validates :data, presence: true` in the base class but with a conditional, like `validates :data, presence: true, unless: skip_validation?`, where `skip_validation?` returns false by default? That way, further notifications (it seems to me, that all but the Badge Notification need the validation) would not need to remember to have `validates :data, presence: true` but those that don't need it have to intentionally disable it.

**Comment by wouterh** [comment]
I will implement something like that tomorrow. Thanks for the feedback.

Op ma 8 feb. 2016 20:35 schreef Anton Rieder <notifications@github.com>:

**Comment by wouterh** [comment]
@aried3r I have implemented your suggestion. I have called the check skip_data_validation? as it causes only the validation of :data to be skipped. The uri etc are still validated.

File: lib/rpush/daemon/wns/toast_request.rb [hunk 1]
```diff
@@ -22,16 +22,23 @@ class ToastRequestPayload
...
           @title = notification.data['title'] || ''
           @body = notification.data['body'] || ''
+          @launch = notification.data['launch']
+          @sound = notification.sound unless notification.sound.eql?("default".freeze)
         end
 
         def to_xml
-          "<toast>
+          launch_string = "" unless @launch
+          launch_string = " launch='#{CleanParamString.clean(@launch)}'" if @launch
```
Comment (aried3r) [inline lib/rpush/daemon/wns/toast_request.rb]: Do you need the extra space before `launch`?

**Comment by aried3r** [comment]
Looking good, just waiting for your answer regarding that extra space.

Also, could you squash the commits into one, once done? :)

File: lib/rpush/daemon/wns/toast_request.rb (hunk 1 above)
Comment (wouterh) [inline lib/rpush/daemon/wns/toast_request.rb]: Yes, it is required. It is used a couple of lines below in:
```
"<toast#{launch_string}>
```

which is meant to produce either:
```xml
<toast>
```
or
```xml
<toast launch='...'>
```

Comment (aried3r) [inline lib/rpush/daemon/wns/toast_request.rb]: Ah, right. I missed that. Thanks for clarification!

**Comment by aried3r** [comment]
Ah, one thing I forgot, would you mind updating the README accordingly to your changes? Would be much appreciated!

**Pull request merged**